engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheHashingJobActor.scala

```diff
@@ -0,0 +1,267 @@
+package cromwell.engine.workflow.lifecycle.execution.callcaching
+
+import java.security.MessageDigest
+import javax.xml.bind.DatatypeConverter
+
+import akka.actor.{ActorRef, LoggingFSM, Props, Terminated}
+import cromwell.backend.standard.callcaching.StandardFileHashingActor.{FileHashResponse, SingleFileHashRequest}
+import cromwell.backend.{BackendInitializationData, BackendJobDescriptor, RuntimeAttributeDefinition}
+import cromwell.core.callcaching.{HashKey, HashResult, HashValue, HashingFailedMessage}
+import cromwell.core.simpleton.WdlValueSimpleton
+import cromwell.engine.workflow.lifecycle.execution.callcaching.CallCacheHashingJobActor.CallCacheHashingJobActorData._
+import cromwell.engine.workflow.lifecycle.execution.callcaching.CallCacheHashingJobActor._
+import wdl4s.values.WdlFile
+
+/**
+  * Actor responsible for calculating individual as well as aggregated hashes for a job.
+  * First calculate the initial hashes (individual and aggregated), and send them to its parent
+  * as well as the callCacheReadingJobActor if one was provided.
+  * From there, either wait for a NextBatchOfFileHashesRequest message from the callCacheReadingJobActor to trigger the next
+  * batch of file hash requests, or send it itself if there is no callCacheReadingJobActor.
+  * When we get all the file hashes for one batch, send those hashes to the CCRead actor and wait for the next NextBatchOfFileHashesRequest.
+  * If it was the last batch and all the files have been hashed, send all the hashes along with the aggregated file hash.
+  * If at any point the callCacheReadingJobActor dies (either because it decided this was a CacheMiss or it found a successful
+  * CacheHit), either keep hashing the files if writeToCache is true, or die if it's not.
+  */
+class CallCacheHashingJobActor(jobDescriptor: BackendJobDescriptor,
+                               callCacheReadingJobActor: Option[ActorRef],
+                               initializationData: Option[BackendInitializationData],
+                               runtimeAttributeDefinitions: Set[RuntimeAttributeDefinition],
+                               backendName: String,
+                               fileHashingActorProps: Props,
+                               writeToCache: Boolean
+                              ) extends LoggingFSM[CallCacheHashingJobActorState, CallCacheHashingJobActorData] {
+
+  val fileHashingActor = makeFileHashingActor()
+
+  // Watch the read actor, as it will die when it's done (cache miss or successful cache hit)
+  // When that happens we want to either stop if writeToCache is false, or keep going
+  callCacheReadingJobActor map context.watch
+
+  initializeCCHJA()
```

Comment:
yeah that's what I thought too but `initialize` is an `FSM` public method so akka isn't happy if you create your own.